### ISSUE TITLE
planar3/iso3 (johndpope fork): sampler assertion crash llama-sampler.cpp:1094

### ISSUE THREAD

**Issue by sztlink** [TheTom/llama-cpp-turboquant#123]
## planar3/iso3 crash from johndpope/feature/planarquant-kv-cache

Testing `johndpope/llama-cpp-turboquant` branch `feature/planarquant-kv-cache` (build `b1-fc3d1b6`) on RTX 4090, all planar3 and iso3 runs crash in the sampler.

**Repro:**
```bash
llama-cli.exe -m llama3.1-8b-q4km.gguf -ngl 99 -ctk planar3 -ctv planar3 -fa 1 -n 256 -c 512 -f prompt.txt
# or: -ctk iso3 -ctv iso3
```

**Error (all context sizes: 512, 16384, 32768):**
```
Assertion failed: found, file ...\llama-sampler.cpp, line 1094
```

Crash after model load + 1–4 tokens generated. q8_0 and turbo3 run fine on the same binary.

**TG speed (RTX 4090, Llama 3.1-8B Q4_K_M, -fa 1 -n 256):**

| Format | Prefill d0 | Gen d0 | Prefill 16K | Gen 16K | Prefill 32K | Gen 32K |
|--------|-----------|--------|------------|---------|------------|---------|
| q8_0   | 1207 t/s  | 122 t/s | 6826 t/s  | 82 t/s  | 7302 t/s   | 63 t/s  |
| turbo3 | 745 t/s   | 102 t/s | **8674 t/s** (+27%) | 50 t/s | 7124 t/s | 44 t/s |
| planar3 | CRASH | — | CRASH | — | CRASH | — |
| iso3   | CRASH | — | CRASH | — | CRASH | — |

The sampler assertion at line 1094 in `llama-sampler.cpp` is checking `found` — with the rotation-based formats, a sampled token ID appears to be failing the vocabulary lookup. Not present with q8_0/turbo3 under the same build.

Pinging @johndpope if you have time to look.

**Comment by TheTom** [comment]
thanks for the detailed bench. q8_0 and turbo3 fine on same binary, so the planar3/iso3 crash is isolated to the rotation kernel in @johndpope's branch... not the sampler itself.

the assert at llama-sampler.cpp:1094 (`found`) fires when sampler tries to look up a token it just sampled. sampler has no contact with KV format, so the path is almost certainly: rotation kernel emits NaN/Inf in attention output → garbage logits → argmax returns out-of-range or repeat-penalty's token-set lookup fails → assert.

quick check that would localize it: dump logits for the first 1–4 sampled tokens. NaN/Inf in there = rotation kernel. bounded but wrong logits = different bug.

planar3/iso3 are johndpope's formats so this one's his to fix, but happy to help triage if a logits dump comes back clean and it's something stranger.
